nessie/src/main/java/org/apache/iceberg/nessie/TableReference.java

```diff
@@ -64,31 +87,137 @@ public static TableReference parse(TableIdentifier path) {
    */
   public static TableReference parse(String path) {
     // I am assuming tables can't have @ or # symbols
-    if (path.split("@").length > 2) {
+    if (REF.splitToList(path).size() > 2) {
       throw new IllegalArgumentException(String.format("Can only reference one branch in %s", path));
     }
-    if (path.split("#").length > 2) {
+    if (TIMESTAMP.splitToList(path).size() > 2) {
       throw new IllegalArgumentException(String.format("Can only reference one timestamp in %s", path));
     }
 
     if (path.contains("@") && path.contains("#")) {
-      throw new IllegalArgumentException("Invalid table name:" +
-          " # is not allowed (reference by timestamp is not supported)");
+      List<String> tableRef = REF.splitToList(path);
+      if (tableRef.get(0).contains("#")) {
+        throw new IllegalArgumentException("Invalid table name:" +
+            " # is not allowed before @. Correct format is table@ref#timestamp");
+      }
+      TableIdentifier identifier = TableIdentifier.parse(tableRef.get(0));
+      List<String> timestampRef = TIMESTAMP.splitToList(tableRef.get(1));
+      return new TableReference(identifier, parseTimestamp(timestampRef.get(1)), timestampRef.get(0));
     }
 
     if (path.contains("@")) {
-      String[] tableRef = path.split("@");
-      TableIdentifier identifier = TableIdentifier.parse(tableRef[0]);
-      return new TableReference(identifier, null, tableRef[1]);
+      List<String> tableRef = REF.splitToList(path);
+      TableIdentifier identifier = TableIdentifier.parse(tableRef.get(0));
+      return new TableReference(identifier, null, tableRef.get(1));
     }
 
     if (path.contains("#")) {
-      throw new IllegalArgumentException("Invalid table name:" +
-          " # is not allowed (reference by timestamp is not supported)");
+      List<String> tableTimestamp = TIMESTAMP.splitToList(path);
+      TableIdentifier identifier = TableIdentifier.parse(tableTimestamp.get(0));
+      return new TableReference(identifier, parseTimestamp(tableTimestamp.get(1)), null);
     }
 
     TableIdentifier identifier = TableIdentifier.parse(path);
 
     return new TableReference(identifier, null, null);
   }
+
+  private enum FormatOptions {
+    DATE_TIME(DateTimeFormatter.ISO_DATE_TIME, ZonedDateTime::from),
+    LOCAL_DATE_TIME(DateTimeFormatter.ISO_LOCAL_DATE_TIME, t -> LocalDateTime.from(t).atZone(sparkTimezoneOrUTC())),
```

Comment:
Looks like we are relying heavily on Spark functionalities through reflection. I am not very familiar with the Nessie module, is this a common pattern? I feel we should make it more generic for other runtime environments.

Reply:
Hey Jack, I am not such a huge fan either. I would prefer to use UTC only. Or in a pinch the system timezone. The Spark check is there to respect the case whena  user has set the spark sql timezone. I haven't found the equivalent parameter in Flink but either way it gets a bit sticky to check via reflection for all engines potential paramters.

I think there are a few options:
1) leave as is and only suppport Spark for the time being
2) remove this and add a parameter/option to set this catalogs timezone
3) only use UTC or system settings.

What do you think?

Reply:
I am fine with the current implementation with Spark specific logic and your refactored code, if that is a primary use case that you need to cover this is probably the best way to have it. But we should definitely take a note for this and think about how interactions and dependencies between different modules, especially with different engines, should be handled in the future.